### ISSUE TITLE
Unclear difference between dependency and package threats

### ISSUE THREAD

**Issue by NVolcz** [slsa-framework/slsa#1039]
The difference between "(D): Use of compromised dependency" and "(H): Use of compromised package" is unclear.
There is no explanation of what make using a package as a dependency different from using it in any other form or way.

I suggest that we either clarify the difference or merge them into one threat.

https://slsa.dev/spec/v1.0/threats-overview
![image](https://github.com/slsa-framework/slsa/assets/1667690/9524d651-3efa-4dfa-bc18-640dfdc37736)

Below are some parts out of a [conversation on Slack](https://openssf.slack.com/archives/C029E4N3DPF/p1710164125448349):

arewm:
> The presence of signed provenance, trusting the signing entity, and its reference to a specific package (i.e. by immutable reference) can be used. Either during consumption of the generated package or consumption of dependencies for the build

nvolcz:
> Thanks! That makes sense! It sound like it is the same threat but mitigated at different times in the process (edited) @

arewm:
> > but mitigated at different times in the process

> Yes, exactly. The threat model has to have a specific reference perspective. It is presented from the lifecycle of a single package/artifact including its dependencies and its consumers. The same rules can apply to both, but they are different from the perspective of a given artifact. (H) would always exist unless you generate a package/artifact which is never used. (H) for one artifact is (D) for its consumer.

**Comment by MarkLodato** [comment]
Agreed on the confusion. Thanks for pointing it out. We should address this.

(D) is really all of the other threats (A)-(H) recursively, from a different frame of reference. It is saying that a given package could be compromised by doing (A)-(H) from one of its dependencies.

Maybe an example would help. Consider [`event-stream`](https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/), which was compromised through its dependency `flatmap-stream`. The attacker snuck in code into the NPM release of `flatmap-stream` that was not present in the source on GitHub. So that's (F) for `flatmap-stream` and (D) for `event-stream`. Does that help?

**Comment by MarkLodato** [comment]
I also agree that the name of (H) could be improved. We originally said "Trick user into using bad package", which sounded limited to typosquatting, but we really also intend things like modification in transit (e.g. MitM).

**Comment by laurentsimon** [comment]
I also think some of the build threats are not relevant where they should be. For example, imo F and G should be under a "publication threat" rather than build threat.

**Comment by MarkLodato** [comment]
I just sent out #1046. Please take a look to see if that's on the right track. I don't think it fully addresses this issue since it doesn't touch (H), but I wanted to keep the PR a manageable size.

**Comment by NVolcz** [comment]
> I don't think so. The idea is that we have three clusters of threats:
> 
>     * "Build threats" = the package that you receive does not match the source of truth
>     * "Source threats" = the source of truth contains unauthorized code
>     * "Dependency threats" = a dependency is compromised, which in turn compromises this one

Your description of the threat clusters is clear and helpful. However, I believe we may need to clarify how to categorize threats that fall at the edges between these clusters.

For instance, consider the threat (C) "Build from modified source." This scenario could be interpreted as a build threat since the package received does not match the source of truth. However, it's not necessarily that the source of truth itself is compromised. So, the question arises: is it a source or a build threat?

Similarly, threats (H) and (D) are positioned at the edge between the group of build threats and consumption of packages. When intercepting the download of a package, is it a build threat or does it a problem for the consumer of the package?

I believe addressing these edge cases will provide a clearer understanding of how to categorize threats within our framework.

**Comment by MarkLodato** [comment]
> There is so much more to publication than build, including delegation of this verification policy (VSA), policy definition, etc...

Yes, all of this is part of the "build track" and "build threats". Perhaps it's just a naming issue. Would "binary provenance threats" or "build & packaging threats" or "publication threats" resonate better?

> I would find it more intuitive if source threats were addressed by source track, build threats by build track, etc

That is the intent. All of these threats are intended to be solved by the build track.

The main challenge is the verification story. Right now the "requirements" for build track are just about publishing provenance and we have this awkward separate page about verification. I want that to all be put together so that the entire track is about end-to-end guarantees. But that's a larger issue.

> For instance, consider the threat (C) "Build from modified source." This scenario could be interpreted as a build threat since the package received does not match the source of truth. However, it's not necessarily that the source of truth itself is compromised. So, the question arises: is it a source or a build threat?

Yes, I completely agree! (C) can go either way:

- We originally had it as part of "build threats" for the reason you said (matching source of truth).
- For v1.0  (#732), we moved it to "source threats" to refocus "build threats" to just be about reporting accurate provenance, and "source threats" kind of being anything about source. But I agree that it's awkward and I don't love it.

> I believe addressing these edge cases will provide a clearer understanding of how to categorize threats within our framework.

👍👍👍

**Comment by kpk47** [comment]
> > There is so much more to publication than build, including delegation of this verification policy (VSA), policy definition, etc...
> 
> Yes, all of this is part of the "build track" and "build threats". Perhaps it's just a naming issue. Would "binary provenance threats" or "build & packaging threats" or "publication threats" resonate better?
> 

This comment feels related to #1041. I don't have a good solution since more precise names tend to be longer and a bit awkward. FWIW, I like "binary provenance threats" best of the listed options.

> > For instance, consider the threat (C) "Build from modified source." This scenario could be interpreted as a build threat since the package received does not match the source of truth. However, it's not necessarily that the source of truth itself is compromised. So, the question arises: is it a source or a build threat?
> 
> Yes, I completely agree! (C) can go either way:
> 
>     * We originally had it as part of "build threats" for the reason you said (matching source of truth).
> 
>     * For v1.0  ([Update threats.md for v1.0 #732](https://github.com/slsa-framework/slsa/pull/732)), we moved it to "source threats" to refocus "build threats" to just be about reporting accurate provenance, and "source threats" kind of being anything about source. But I agree that it's awkward and I don't love it.
> 

I see (B) as an attacker making unauthorized changes to the official source and (C) as building from unofficial source. I think we're having trouble classifying (C) because we're assuming some notion of official source that isn't a part of the SLSA model. 

The shape of my thinking is roughly:
- SLSA creates a verifiable link between an artifact and its source repo.
- Package repos maintain a mapping from {package name, version} to artifact.
- _Some_ package repos maintain a mapping from {package name, version} to official source repo. Otherwise nobody does.

(C) is an attack on that last mapping and only makes sense in a context where that mapping is explicit and maintained. In such a context, I might argue that it's an attack on the publication process rather than the build process or source repo.

It also isn't clear to me how much this classification matters since perhaps this ambiguity is evidence that the source/build/dependencies/etc classification is breaking down.

**Comment by marcelamelara** [comment]
@TomHennen @lehors @adriandiglio This issue actually seems related to the changes we've introduced in v1.1. This may be something we need to revisit during the RFC period for v1.1 RC2. Thoughts?

**Comment by arewm** [comment]
#1046 was meant to at least partially address this issue but there was a comment
> More work is needed to refine (H), but that
is left to a separate PR.

and that this issue
> mentions confusion between the distinction of a consumer as something potentially separate from a dependency relationship.
>
> We can make it more apparent that a dependency is a form of a consumer relationship.

I feel like this is verging on going in the direction of policies for consuming/verifying. We start to broach policies with the source track. Can we defer further refinement/resolution on this issue until after v1.1?

**Comment by TomHennen** [comment]
I'd argue that this _has_ been addressed in the RC and we can close.


It now includes [Distribution channel](https://slsa.dev/spec/v1.1-rc2/threats#i-usage:~:text=(G)%20Distribution%20channel,-An%20adversary%20modifies) threats and their migtations which is upstream of both the consumer and the dependnecy.

In addition we now explicitly call out that ["This version of SLSA does not explicitly address dependency threats"](https://slsa.dev/spec/v1.1-rc2/threats#dependency-threats:~:text=This%20version%20of%20SLSA%20does%20not%20explicitly%20address%20dependency%20threats).

So I think we're good.
